src/arch/Sound/RageSoundDriver_Generic_Software.cpp

```diff
@@ -476,69 +472,54 @@ RageSoundDriver::~RageSoundDriver()
+	// the maximum hardware frame is updated.
+	//
+	// Otherwise, the maximum hardware frame is returned.
+	// The value of p represents the power of 2 of the maximum hardware frame.
+	// This is because stereo audio is processed in blocks of 2^p frames.
+
+	const static std::int64_t one = 1;
+	const static std::int64_t pInitial = 22;
+	static std::int64_t lastLogTime = 0;
+
+	if (iHardwareFrame > m_iMaxHardwareFrame)
+	{
+		m_iMaxHardwareFrame = iHardwareFrame;
+	}
 
-	//iHardwareFrame %= 0x800000; // debug test a sample value max of about 3 minutes so I dont have to spend an hour per test
-	std::int64_t diff = iHardwareFrame - m_iMaxHardwareFrame;
-	if( diff < 0 )
+	std::int64_t currentTime = RageTimer::deltaMicrosecondsAsSigned();
+	if (lastLogTime == 0 || (currentTime - lastLogTime) > 1000000)
 	{
-		diff = 0;
-		int iMinuteSampleRate = GetSampleRate()*60; //get one minute worth of grace -- if you need more, there is very likely some other problem going on
-		//if we have a sample clamp and the new hardware frame is within a fresh minute of the sample rate max and have 'underrun'
-		if ((soundDriverMaxSamples>0) && (iHardwareFrame<iMinuteSampleRate) && iHardwareFrame >= 0)
-		{
-			LOG->Trace("RageSoundDriver: driver position mask adjustment hardware frame number: %d, last max frame number: %d, soundDriverMaxSamples: %d, iMinuteSampleRate: %d, m_iVMaxHardwareFrame: %d",
-																			(int)iHardwareFrame,    (int)m_iMaxHardwareFrame,  soundDriverMaxSamples,  iMinuteSampleRate, (int) m_iVMaxHardwareFrame);
-			diff = (soundDriverMaxSamples - m_iMaxHardwareFrame) + iHardwareFrame;
-			m_iMaxHardwareFrame = 0;
-		}
-		else
+		// Assuming m_iMaxHardwareFrame is a positive integer and not zero.
+		if (m_iMaxHardwareFrame > 0)
 		{
-			/* Clamp the output to one per second, so one underruns don't cascade due to
-			 * output spam. */
-			static RageTimer last(RageZeroTimer);
-			if (last.IsZero() || last.Ago() > 1.0f)
-			{
+			std::int64_t p = 0;
+			std::int64_t tempFrame = m_iMaxHardwareFrame;
+			while (tempFrame >>= 1) ++p;
 
-				//try to hand hold the user if their audio driver is possibly bad
-				int p = 21; // save some time, assume the buffer has at least a minute of cd quality audio -- 2^21
-				while (std::pow(2,p) < m_iMaxHardwareFrame)
-				{
-					if (p == 31)  break; //do not want to go beyond signed DWORD size
-					p++;
-				}
-
-				LOG->Trace("RageSoundDriver: driver returned a lesser position (%d < %d). If this is a recurrent driver problem with your sound card and not an underrun, try setting the preference RageSoundSampleCountClamp to %d",
-					(int)iHardwareFrame, (int)m_iMaxHardwareFrame, (int)std::floor(std::pow(2.0, p)));
-				last.Touch();
+			// Find the nearest power of 2 to the maximum hardware frame.
+			if ((one << p) < m_iMaxHardwareFrame)
+			{
+				p++;
 			}
 
-			//return m_iMaxHardwareFrame;
+			// Ensure p does not go beyond 31 to avoid exceeding signed DWORD size.
+			if (p > 31) p = 31;
```

Comment:
I don't actually know if we need to make sure the result fits into a signed DWORD.  Considering it's been the standard behavior in SM5 for at least 10 years, I didn't want to remove this logic without being absolutely certain it's not needed. 

I would assume, since it came from a PR focused on fixing DirectSound in Windows XP, that if DirectSound/WaveOut/WDMKS all work without this logic in place, then we don't need it. I wouldn't want to remove it until I'm absolutely sure, though.